### ISSUE TITLE
can't initialize the compacted realm at the second thread

### ISSUE THREAD

**Issue by JosephZZ** [realm/realm-cocoa#3348]
## Goals
At the beginning without any multithreading and any other opened realm instance, i added many entries to realm and then I niled out the instance and compacted the database. 
After that, I need to access to Realm with several threads together. The first thread initiated an instance successfully, but the second thread just hang there at the line `try! realm = Realm()`
>  

## Code Sample

The logic goes like this:
at the start of the app, we initialize the database in the main thread. After it finished, we start to use the existing data to do some more tasks in different threads later.
```Swift
func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {    
            DbInitializerAndUpdater().run()
}
```
In DbInitializerAndUpdater, we have
```Swift
class DbInitializerAndUpdater{
    var realm : Realm? = try! Realm()
    run(){
        addAllPhotoMetaDataFromPhotoLibrary()
        // the following code compacts realm
        let paths = NSSearchPathForDirectoriesInDomains(.LibraryDirectory, .UserDomainMask, true)
        let documentDir = paths[0]
        let compactPath = documentDir + "/randomeStringBlahBlah.realm"
        try! realm!.writeCopyToPath(compactPath)
        let originalPath = realm!.path
        realm = nil
        try! NSFileManager.defaultManager().replaceItemAtURL(NSURL(fileURLWithPath: originalPath),     withItemAtURL: NSURL(fileURLWithPath: compactPath), backupItemName: nil, options: .UsingNewMetadataOnly, resultingItemURL: nil)
     }
}
```

And then later, we have an image processor which is processing photos in background thread, while the another thread is selecting processed images for displaying. 
The image processor is calling a class `PhotoSelectorForImageProcessor`, and the displayin thread is calling a class `PhotoSelectorForDisplaying`, but both class inherit from the same parent class called 'BasePhotoSelector'. They look like:
```Swift
class BasePhotoSelector{
      var realm : Realm!
      init(){
        printDebug("try to initalize realm")
        realm = try! Realm()
        printDebug("got realm")
    }
//all common methods
}

class PhotoSelectorForImageProcessor : BasePhotoSelector{
    override init() {
        super.init()
    }
//some more methods
}

class PhotoSelectorForDisplaying : BasePhotoSelector{
    override init() {
        super.init()
    }
//some more methods
}
```
And then what happened in console is that it'll print out:
```
try to initalize realm
got realm
try to initalize realm
```

The code will just hang at the line ` realm = try! Realm()` and never reach `got realm`.

## Version of Realm and Tooling
0.98.5
> I'm using cocoapods

**Comment by jpsim** [comment]
```
try! NSFileManager.defaultManager().replaceItemAtURL(NSURL(fileURLWithPath: originalPath),     withItemAtURL: NSURL(fileURLWithPath: compactPath), backupItemName: nil, options: .UsingNewMetadataOnly, resultingItemURL: nil)
```

you can't replace the file while Realm is still accessing it. I've written more about this in https://github.com/realm/realm-cocoa/issues/3240#issuecomment-194590128

**Comment by JosephZZ** [comment]
so I changed the line `realm = nil` into `realm!.invalidate()` but it doesn't work. What should I do to make sure no realm is accessing it?

**Comment by jpsim** [comment]
> What should I do to make sure no realm is accessing it?

To do the same with all other Realms/Objects/Results/Lists/etc on all other threads.

**Comment by JosephZZ** [comment]
But the thing is all other threads starts only after the DbInitializerAndUpdater.run() function has finished

**Comment by jpsim** [comment]
@JosephZZ please share a full reproduction case and we'll be able to identify if that's actually the case, or if there's a bug in Realm.

**Comment by JosephZZ** [comment]
so i found out that it's not just about multithread access. I tried to make a simple reproduction with 2 threads accessing at the same time after compacting at the beginning, but it ran with no problem. So I looked closer at my real code base and found out that it should be a dead lock problem. 2 threads both stops at " __psynch_mutexwait ()". But this dead lock only happens if i compacted the database at the beginning (yeah it compacted when no other thread is accessing), but wouldn't happen if I hadn't compact it. When the deadlock happen, one thread is trying to make queries, while the other thread is trying to create a new realm instance.

I'll keep trying to make a reproduction, but let me know if the above new information can help any?

**Comment by JosephZZ** [comment]
https://github.com/JosephZZ/RealmDBCompactionBugReproduction

Here's the bug reproduction. 
Basically, if the db is compacted, it can't be accessed by different threads. The second thread just get stuck at trying to create an instance but never go through. (though in my real code, what happened is that two threads entered into a something like a deadlock - both stopped at "__psynch_mutexwait ()")

**Comment by jpsim** [comment]
The problem in your code sample is that you delete & replace the Realm file while other threads are still using it. This is obviously not something that's supported. However, with a few modifications to your code, you can be sure to not be accessing the file while it's being replaced:

```diff
diff --git a/RealmCompactProblemRepro/AppDelegate.swift b/RealmCompactProblemRepro/AppDelegate.swift
index 13d4de3..fac7145 100644
--- a/RealmCompactProblemRepro/AppDelegate.swift
+++ b/RealmCompactProblemRepro/AppDelegate.swift
@@ -18,7 +18,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         var GlobalBackgroundQueue: dispatch_queue_t {
             return dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0)
         }
-        DbInitializer().run()
+        // wrap Realm initialization & compaction in an autoreleasepool
+        // to force it to be closed afterwards
+        autoreleasepool {
+            DbInitializer().run()
+        }
         dispatch_async(GlobalBackgroundQueue){
             ImageProcessor().startProcessing()
         }
diff --git a/RealmCompactProblemRepro/ViewController.swift b/RealmCompactProblemRepro/ViewController.swift
index d5b93c6..fc48ffb 100644
--- a/RealmCompactProblemRepro/ViewController.swift
+++ b/RealmCompactProblemRepro/ViewController.swift
@@ -10,7 +10,8 @@ import UIKit
 import RealmSwift
 
 class ViewController: UIViewController {
-    let realm = try! Realm()
+    // make realm variable lazy to avoid accessing it before it's compacted
+    lazy var realm = try! Realm()
     override func viewDidLoad() {
         super.viewDidLoad()
         print("start vc")
```

**Comment by JosephZZ** [comment]
it does seem to work! it didn't say that the compacting can be still continuing after the the function returns. And unfortunately i have always assumed so. thanks a lot for the help!

So if now I want to compact db in the middle of running the app, is there a way to release all the realm instances easily? Or i'd better do it at the beginning of the app ?

**Comment by jpsim** [comment]
> it does seem to work! it didn't say that the compacting can be still continuing after the the function returns. And unfortunately i have always assumed so. thanks a lot for the help!

`writeCopyToURL(_:)` is synchronous and returns immediately after the operation completes. What "lingers" here is the strong reference to the initial `Realm` instance, which is just ARC working as designed. The use of `autoreleaspool` above forces the instance to be released when the pool drains, therefore preventing this instance from "lingering".

> So if now I want to compact db in the middle of running the app, is there a way to release all the realm instances easily? Or i'd better do it at the beginning of the app ?

Because Realm database "connections" are tied to the instance lifetimes of `Realm` classes, asking how to release the realm instances "easily" is exactly the same as asking how to release *any* collection of Swift class instances "easily". With ARC, the answer is to essentially use `autoreleasepool`s to drain unreferenced Realm instances. If you have Realm, Realm Objects, or Realm Collections that are strongly referenced by view controllers or others, it will be difficult if not impossible to release all Realm instances.

This is an unavoidable design constraint from tying the lifetimes of Realm "db connections" to the lifetimes of the object instances.

It'll be much simpler and predictable to write a copy and overwrite the initial file at app launch, before any Realm accessors are created.

**Comment by slreddy1137** [comment]
Hi,

I have been trying to use the writeCopyToURL when the app is going to background, this is reducing the Realm file size, but when i re-launch the app there are places where i m creating a Realm instance inside a dispatch queue those are not getting created and ignoring the complete code, and its crashing due to some old references, i have been trying to figure out to fix this but unable to.

can someone guide me with best practise please.

**Comment by jpsim** [comment]
@slreddy1137 please open a new issue, as this ticket is fairly specific to @JosephZZ's exact use case and might not apply to yours. It will also be easier for us to help you if you share more specific information, like a sample project that demonstrates the problems you're facing.

And @JosephZZ, I hope this discussion, my proposed solutions and final explanation have helped you. I'll close this issue because I think we've sufficiently explained what's happening here. If you continue to have problems in the future, please open a new issue and we'll do our best to help again.
